src/libAtomVM/atom_table.c

```diff
@@ -0,0 +1,479 @@
+{
+    SMP_RDLOCK(table);
+
+    struct HNode *node = get_node_using_index(table, index);
+    if (IS_NULL_PTR(node)) {
+        SMP_RDLOCK(table);
+        return NULL;
+    }
+
+    *out_len = atom_string_len(node->key);
+
+    SMP_UNLOCK(table);
+    return node;
+}
+
+void atom_table_write_bytes(struct AtomTable *table, atom_ref_t atom, size_t buf_len, void *outbuf)
+{
+    SMP_RDLOCK(table);
+
+    struct HNode *node = (struct HNode *) atom;
+    size_t len = atom_string_len(node->key);
+    if (len > buf_len) {
+        len = buf_len;
+    }
+
+    memcpy(outbuf, atom_string_data(node->key), len);
+
+    SMP_UNLOCK(table);
+}
+
+void atom_table_write_cstring(
+    struct AtomTable *table, atom_ref_t atom, size_t buf_len, char *outbuf)
+{
+    SMP_RDLOCK(table);
+
+    struct HNode *node = (struct HNode *) atom;
+    atom_string_to_c(node->key, outbuf, buf_len);
+
+    SMP_UNLOCK(table);
+}
+
+static inline void init_node(struct HNode *node, AtomString atom, long index)
+{
+    node->key = atom;
+    node->index = index;
+}
+
+static inline void insert_node_into_bucket(
+    struct AtomTable *table, int bucket_index, struct HNode *node)
+{
+    struct HNode *maybe_existing_node = table->buckets[bucket_index];
+    table->buckets[bucket_index] = node;
+    node->next = maybe_existing_node;
+}
+
+static inline long insert_node(struct AtomTable *table, struct HNodeGroup *node_group,
+    unsigned long bucket_index, AtomString string)
+{
+    long new_index = table->count;
+    table->count++;
+
+    struct HNode *node = &node_group->nodes[new_index - node_group->first_index];
+    table->last_node_group_avail--;
+    init_node(node, string, new_index);
+    insert_node_into_bucket(table, bucket_index, node);
+
+    return new_index;
+}
+
+static void do_rehash(struct AtomTable *table, int new_capacity)
+{
+    int new_size_bytes = sizeof(struct HNode *) * new_capacity;
+    struct HNode **new_buckets = realloc(table->buckets, new_size_bytes);
+    if (IS_NULL_PTR(new_buckets)) {
+        // Allocation failure can be ignored, the hash table will continue with the previous bucket
+        return;
+    }
+    memset(new_buckets, 0, new_size_bytes);
+    table->buckets = new_buckets;
+
+    struct HNodeGroup *group = table->first_node_group;
+
+    while (group) {
+        int group_count;
+        if (group == table->last_node_group) {
+            group_count = group->len - table->last_node_group_avail;
+        } else {
+            group_count = group->len;
+        }
+
+        for (int i = 0; i < group_count; i++) {
+            struct HNode *node = &group->nodes[i];
+            AtomString key = node->key;
+
+            unsigned long hash = sdbm_hash(key, atom_string_len(key));
+            unsigned long bucket_index = hash % table->capacity;
+
+            insert_node_into_bucket(table, bucket_index, node);
+        }
+
+        group = group->next;
+    }
+}
+
+static inline void maybe_rehash(struct AtomTable *table, int new_entries)
+{
```

Comment:
I can't find where table->capacity is updated.

Reply:
Fixed.